### PR TITLE
System properties

### PR DESCRIPTION
**Pull request by rockjam**
This one is work in progress. I'm not sure if it is right way to pass system properties to build tasks and user's code. In https://github.com/rockjam/cbt/blob/system-properties/examples/system-properties/README.md I described how it works in current implementation.

### CODE REVIEW

File: stage2/Stage2.scala [hunk 1]
```diff
@@ -46,6 +48,10 @@ object Stage2 extends Stage2Base{
       args.compatibilityTarget,
       null
     )
+
+    // set System.properties passed as argument to task.
+    context.propsMap().asScala foreach { case (k, v) => System.setProperty(k, v) }
```
Comment (rockjam) [inline stage2/Stage2.scala]: I really don't know where is the best place for it

**Comment by cvogt** [comment]
Very interesting.

This PR seems to achieve the same thing for the nailgun client: https://github.com/martylamb/nailgun/pull/17 but it is not merged. Possibly because system properties are global mutable stage and multiple clients would interfere with each other.

We have the same problem with CBT, so it is not advisable to use system properties with CBT. At the same time we should have a mechanism to provide them for things that already rely on them.

I wonder if there is some way to highjack the system property mechanism in the jvm to make reject mistakes (e.g. reject thread writing or reading system properties other still active threads have set) or fix the behavior by scoping them to threads via thread locals and possibly copying the local scope when a thread creates a new thread. (Only slightly related: we might have to consider similar crazy hackery to make System.console work).

I am supportive with this PR, I need more time to review it :). Good stuff @rockjam

**Comment by cvogt** [comment]
The before-mentioned high-jacking may involve using the SecurityManager to detect system property reads and writes and/or byte-code rewriting to highjack the mechanism.

**Comment by cvogt** [comment]
@rockjam FYI, this PR fails because the change to ContextImplementation broke source compatibility. The way to sent another commit on top of this that replaces references to old versions with the previous commit of this PR. In particular search for the two occurences of `1f4f6097d3ca682d6fd20a7cc6dd277832350827` and replace it with `f57386c2ba95bf47309f930eeaa48fd4d88368ab`. But let's delay this until review is done, because if it involves rebasing the commit hashes will change.

**Comment by cvogt** [comment]
@rockjam sorry for the delay. So I have thought about this. One thing I don't like about this yet is that system properties are global mutable state, which permanently affects the background process when people use nailgun. I think people should ideally be conscious that they are affecting global mutable state shared by all cbt instances running, but a persistent background process is unusual for many people.

I know you are using cbt without nailgun, so you are not affected by this, but system properties don't work well with nailgun for that reason. So how about we do this:

We separate the mechanism to affect system properties from the mechanism we provide properties to cbt. We have two mechanisms. CBT's own mechanism is not global but scoped to the cbt instance and disappears when it terminates eventhough the nailgun background process keeps running. We also provide a separate mechanism to affect system properties but urge people to only use it when they have to affect system properties out of their control and warn them about the implications.

In detail this could look like this:

`cbt -Xlog=all -Dscala.color=true repl`

`-X` are the cbt specific properties that are NOT turned into system properties (the name is borrowed from the java/scala equivalents). `-Xlog` sets the loggers, just like`-Dlog` currently, independent for every cbt run, no global mutable state. `-Dscala.color=true` will actually set a system property. However cbt exits with an error when trying to use `-D` when running on nailgun. So you can't use it. Instead you'll have to do this:

`cbt setProperties -Dscala.color=true` which returns an error when NOT running on nailgun, but sets the system property otherwise.

This highlights very well that system properties are performing mutation.

If we want a way to have the exact same cbt command line invocation work with and without nailgun, we can consider an additional flag, e.g. `-Xunsafe` to raise awareness of the risk.

`cbt -Xlog=all -Xunsafe -Dscala.color=true repl`

This would then work regardless of nailgun or not and set the system properties.

Also, we should probably only allow `-X` and `-D` to be provided BEFORE task name, so 

`cbt <allowed here> <task name> <but not here>`

so that the task could itself take -X or -D arguments, e.g. if it is the Scala repl, etc.

WDYT?

**Comment by cvogt** [comment]
now I am thinking maybe we should remove the `X` from regular cbt arguments, so just `-log=all` and `-unsafe`. And we should not pass things like the value of `log` through in an untyped map. 
But I can see having `-X...` args in an untyped map for user extensibility.

**Comment by rockjam** [comment]
@cvogt
System you described seems difficult to me. There are some things I agree, but want to make it simpler somehow, and I have some suggestions.

1. I agree to separate cbt's properties from system properties, so we could do:

  ```
  cbt -log=all -debug=true repl -Dscala.color=true
  ```
  where `-log` and `-debug` flags are cbt's, and `-Dscala.color=true` is global flag passed to JVM

2. Don't let people use system properties with nailgun at all.
  We already have warnings in cbt, saying: "use cbt direct instead". We could do the same in cases when user need to pass system properties.
  So, allow only:
  ```
  cbt direct repl -Dscala.color=true
  ```
  but not:
  ```
  cbt repl -Dscala.color=true
  ```
3. Don't let people to pass system properties to task itself.
  > Also, we should probably only allow -X and -D to be provided BEFORE task name, so
  > ```cbt <allowed here> <task name> <but not here>```
  > so that the task could itself take -X or -D arguments, e.g. if it is the Scala repl, etc.
  
  Want to find out, if we are on the same page. AFAIK, we just pass system properties globally, to instance of JVM that is running current task. System properties will be the same inside cbt's task and inside the thing that task may run. For example, when we execute `cbt run -Duser.timezone=UTC`, we'll get same value of `System.getProperty("user.timezone")` inside `run` task and inside `Main` class. This is how it works.
  
  I also think that we could avoid passing system properties to tasks in favour args(#172) of environment variables.

**Comment by cvogt** [comment]
> 1.

Sounds good!

> Don't let people use system properties with nailgun at all.

Good idea. Let's start with that. We can always start supporting it in the future if there is need, but let's keep it simple for now. 👍 

> I also think that we could avoid passing system properties to tasks in favour args(#172) of environment variables.

I think we need to support it for interaction with existing libraries like logback.

> 3.

I just tried this:

```
$ echo 'println(System.getProperty("foo") + args)' > test.scala
$ scala -Dfoo="hello" test.scala asdf
hello WrappedArray(asdf)
$ scala test.scala asdf -Dfoo="hello"
hello WrappedArray(asdf, -Dfoo=hello)
```

So interestingly Scala turns -D into system properties AND passes them as arguments to the application. Java handles it differently:

```
$ echo 'object Foo extends App{ println(System.getProperty("foo") + " " + args.toSeq) }' > test.scala
$ scalac test.scala asdf
$ java -cp (cbt tools resolve org.scala-lang:scala-library:2.10.6):. -Dfoo="hello" Foo asdf
hello WrappedArray(asdf)
$ java -cp (cbt tools resolve org.scala-lang:scala-library:2.10.6):. Foo asdf -Dfoo="hello"
null WrappedArray(asdf, -Dfoo=hello)
```

So Java only interprets them as system properties when passed BEFORE the name of the application.

I feel `java` acts more reasonably here. With `scala` as seen above there is no way to pass `-Dfoo=hello` to a program without making it a system property. An application (or task in cbt land) may however require a `-D` argument for some reason, e.g. because it passes arguments to a non-java command line tool like `ssh`, where `ssh -D...` has a different meaning. `cbt` currently prevents this. `scala` seems to entangle it with system properties. `java` seems to be doing the right thing and we should probably make `cbt` do the same thing.

**Comment by cvogt** [comment]
moving to milestone 1.0, since this would clash too much with the planned refactorings afterwards

**Pull request closed without merging**